### PR TITLE
Colshape type

### PR DESCRIPTION
**Pull request by CrosRoad95**
**Function name:** `getColShapeType`
**Type:** Shared
**Method:** `Colshape:getShapeType( )`
**Variable:** `shapeType`
**Syntax:** `string shapeType getColShapeType( colshape shape )`

shape type can be: Circle, Cubboid, Sphere, Rectangle, Polygon, Tube

Test script type: shared
```lua
local circle = createColCircle(0,0,1)
local cubboid = createColCuboid(0,0,0,0,0,0)
local sphere = createColSphere(0,0,0,0)
local rectangle = createColRectangle(0,0,0,0)
local polygon = createColPolygon(0,0,0,0,0,0,0,0)
local tube = createColTube(0,0,0,0,0)

iprint("circle",getColShapeType(circle), circle:getShapeType(), circle.shapeType)
iprint("cubboid",getColShapeType(cubboid), cubboid:getShapeType(), cubboid.shapeType)
iprint("sphere",getColShapeType(sphere), sphere:getShapeType(), sphere.shapeType)
iprint("rectangle",getColShapeType(rectangle), rectangle:getShapeType(), rectangle.shapeType)
iprint("polygon",getColShapeType(polygon), polygon:getShapeType(), polygon.shapeType)
iprint("tube",getColShapeType(tube), tube:getShapeType(), tube.shapeType)
```

### CODE REVIEW

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 1]
```diff
@@ -36,6 +36,7 @@ void CLuaElementDefs::LoadFunctions ( void )
     CLuaCFunctions::AddFunction ( "isElementWithinColShape", IsElementWithinColShape );
     CLuaCFunctions::AddFunction ( "isElementWithinMarker", IsElementWithinMarker );
     CLuaCFunctions::AddFunction ( "getElementsWithinColShape", GetElementsWithinColShape );
+    CLuaCFunctions::AddFunction ( "getColShapeType", GetColShapeType );
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: Should be in `CLuaColShapeDefs::LoadFunctions`

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 2]
```diff
@@ -891,6 +892,30 @@ int CLuaElementDefs::IsElementWithinMarker ( lua_State* luaVM )
     return 1;
 }
 
+int CLuaElementDefs::GetColShapeType ( lua_State* luaVM )
+{
+    // Verify the arguments
+    CClientColShape* pColShape = NULL;
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: Use `nullptr`

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 3]
```diff
@@ -891,6 +892,30 @@ int CLuaElementDefs::IsElementWithinMarker ( lua_State* luaVM )
...
+    CScriptArgReader argStream ( luaVM );
+    argStream.ReadUserData ( pColShape );
+
+    if ( ! argStream.HasErrors ( ) )
+    {
+        // Grab our VM
+        CLuaMain* pLuaMain = m_pLuaManager->GetVirtualMachine ( luaVM );
+        if (pLuaMain)
+        {
+            lua_pushnumber ( luaVM, pColShape->GetShapeType ( ) + 1 );
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: Any specific reason not to use the 0 value here?

File: Server/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.cpp [hunk 4]
```diff
@@ -34,6 +34,10 @@ void CLuaColShapeDefs::AddClass ( lua_State* luaVM )
     lua_classfunction ( luaVM, "Polygon", "createColPolygon" );
 
     lua_classfunction ( luaVM, "getElementsWithin", "getElementsWithinColShape" );
+    lua_classfunction ( luaVM, "getShapeType", "getColShapeType" );
+
+    lua_classvariable ( luaVM, "shape", NULL, "getColShapeType" );
```
Comment (Necktrox) [inline Server/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.cpp]: use `nullptr`

File: Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 5]
```diff
@@ -47,6 +47,7 @@ void CLuaElementDefs::LoadFunctions ( void )
     CLuaCFunctions::AddFunction ( "getElementType", getElementType );
     CLuaCFunctions::AddFunction ( "getElementInterior", getElementInterior );
     CLuaCFunctions::AddFunction ( "getElementsWithinColShape", getElementsWithinColShape );
+    CLuaCFunctions::AddFunction ( "getColShapeType", getColShapeType );
```
Comment (Necktrox) [inline Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: Should be in `CLuaColShapeDefs::LoadFunctions`

File: Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 6]
```diff
@@ -964,6 +965,30 @@ int CLuaElementDefs::isElementWithinColShape ( lua_State* luaVM )
     return 1;
 }
 
+int CLuaElementDefs::getColShapeType ( lua_State* luaVM )
```
Comment (Necktrox) [inline Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: Move to `/Server/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.cpp`

File: Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 7]
```diff
@@ -964,6 +965,30 @@ int CLuaElementDefs::isElementWithinColShape ( lua_State* luaVM )
     return 1;
 }
 
+int CLuaElementDefs::getColShapeType ( lua_State* luaVM )
+{
+    // Verify the arguments
+    CColShape* pColShape = NULL;
```
Comment (Necktrox) [inline Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: Use `nullptr`

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.h [hunk 8]
```diff
@@ -38,6 +38,7 @@ class CLuaElementDefs : public CLuaDefs
     LUA_DECLARE ( IsElementWithinColShape );
     LUA_DECLARE ( IsElementWithinMarker );
     LUA_DECLARE ( GetElementsWithinColShape );
+    LUA_DECLARE ( GetColShapeType );
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.h]: Move to `CLuaColShapeDefs.h`

File: Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.h [hunk 9]
```diff
@@ -47,6 +47,7 @@ class CLuaElementDefs: public CLuaDefs
     LUA_DECLARE ( getElementsByType  );
     LUA_DECLARE ( getElementInterior );
     LUA_DECLARE ( getElementsWithinColShape );
+    LUA_DECLARE ( getColShapeType );
```
Comment (Necktrox) [inline Server/mods/deathmatch/logic/luadefs/CLuaElementDefs.h]: Move to `CLuaColShapeDefs.h`

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp [hunk 10]
```diff
@@ -891,6 +892,30 @@ int CLuaElementDefs::IsElementWithinMarker ( lua_State* luaVM )
     return 1;
 }
 
+int CLuaElementDefs::GetColShapeType ( lua_State* luaVM )
```
Comment (Necktrox) [inline Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: Move to `/Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.cpp`

**Review by Necktrox — CHANGES_REQUESTED** [review]

File: Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp (hunk 3 above)
Comment (CrosRoad95) [reply Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: becose lua start index from 1, not 0 like c++ and other languages

Comment (Necktrox) [reply Client/mods/deathmatch/logic/luadefs/CLuaElementDefs.cpp]: But in this case it's not an index

File: Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.cpp [hunk 11]
```diff
@@ -33,11 +35,39 @@ void CLuaColShapeDefs::AddClass ( lua_State* luaVM )
     lua_classfunction ( luaVM, "Polygon", "createColPolygon" );
 
     lua_classfunction ( luaVM, "getElementsWithin", "getElementsWithinColShape" );
+    lua_classfunction ( luaVM, "getShapeType", "getColShapeType" );
+
     lua_classvariable ( luaVM, "elementsWithin", NULL, "getElementsWithinColShape" );
+    lua_classvariable ( luaVM, "shape", NULL, "getColShapeType" );
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/luadefs/CLuaColShapeDefs.cpp]: the attribute should be `shapeType` here

**Review by qaisjp — CHANGES_REQUESTED** [review]
The function should return a string enum.

**Review by qaisjp — APPROVED** [review]

**Review by qaisjp — COMMENTED** [review]
Dismissing review

**Comment by qaisjp** [comment]
It returns an integer now, starting with 0

**Pull request merged**